network/framework/src/connectivity_manager/mod.rs

```diff
@@ -482,47 +525,47 @@ where
-            let stale_peers = self
-                .connected
-                .iter()
-                .filter(|(peer_id, _)| !trusted_peers.contains_key(peer_id))
-                .filter_map(|(peer_id, metadata)| {
-                    // If we're using server only auth, we need to not evict unknown peers
-                    // TODO: We should prevent `Unknown` from discovery sources
-                    if !self.mutual_authentication
-                        && metadata.origin == ConnectionOrigin::Inbound
-                        && (metadata.role == PeerRole::ValidatorFullNode
-                            || metadata.role == PeerRole::Unknown)
-                    {
-                        None
-                    } else {
-                        Some(*peer_id) // The peer is stale
-                    }
-                });
-
-            // Close existing connections to stale peers
-            for stale_peer in stale_peers {
-                info!(
-                    NetworkSchema::new(&self.network_context).remote_peer(&stale_peer),
-                    "{} Closing stale connection to peer {}",
-                    self.network_context,
-                    stale_peer.short_str()
-                );
+            let trusted_peers = trusted_peers.read().clone();
+            for (peer_network_id, metadata) in self.peer_metadata_cache.iter() {
+                if trusted_peers.contains_key(&peer_network_id.peer_id()) {
+                    continue; // trusted is never stale
+                }
+                if !self.config.mutual_authentication
+                    && metadata.connection_metadata.origin == ConnectionOrigin::Inbound
+                    && (metadata.connection_metadata.role == PeerRole::ValidatorFullNode || metadata.connection_metadata.role == PeerRole::Unknown) {
+                    continue; // not stale
+                }
+                // is stale! Close...
 
-                if let Err(disconnect_error) =
-                    self.connection_reqs_tx.disconnect_peer(stale_peer).await
-                {
-                    info!(
-                        NetworkSchema::new(&self.network_context)
-                            .remote_peer(&stale_peer),
-                        error = %disconnect_error,
-                        "{} Failed to close stale connection to peer {}, error: {}",
-                        self.network_context,
-                        stale_peer.short_str(),
-                        disconnect_error
-                    );
+                match self.peer_senders.get_generational(self.peer_senders_generation) {
+                    None => {}
+                    Some((new_peer_senders, new_generation)) => {
+                        self.peer_senders_cache = new_peer_senders;
+                        self.peer_senders_generation = new_generation;
+                    }
+                }
+                #[cfg(disabled)] // TODO: actually closing 'stale' is disabled until fixed
```

Comment:
Were there only specific types of connections being closed? Or all of them? My first guess is that somewhere the roles might have been switched up 🤔